### ISSUE TITLE
Add ability to define possible values for query params

### ISSUE THREAD

**Issue by svenluijten** [usebruno/bruno#344]
For example, we support a `sort` query param in our API that can have the following values: 

- `start`
- `-start`
- `end`
- `-end`

It would be great if I could set the "possible values" for the `sort` query parameter in Bruno, so I don't have to go back to the code/documentation to check what options I have.

**Comment by helloanoop** [comment]
> Did we miss something?

Currently Bruno does not have this. I really like the idea.

Here is how a rough idea of how we can fit this in Bruno's way of doing things.
Open to feedback and changing this approach.

We store a the choices in the request vars using `@choice()` annotation
![image](https://github.com/usebruno/bruno/assets/9350733/9dc097be-7c21-4329-a268-193a4e6794f4)

If a choice variable gets referenced inside a query param, then we should a select dropdown in the right side of the value cell.
![image](https://github.com/usebruno/bruno/assets/9350733/ec641047-ed33-4762-a3f1-02da94819797)

@ajaishankar Do you have any in inputs on the Bru Lang DSL ? 

> Delivery Timeline

A bit tied with some things. If we agree on the design then I don't think any Bru Lang Transpiler updates would be needed.
This should be just a UI level change.

**Comment by conejoninja** [comment]
In case of documentation/params explanation, it just need to show up as a different tab, content could be in markdown.

For example response, a json/xml would be enough in most cases, but I guess the same as before applies, just a new tab/area to show some markdown content.

Just like Postman do, this would be great not only for Bruno's UI, but for other form of representation (like a website)

**Comment by svenluijten** [comment]
I like the idea of a `@choice()` annotation to facilitate this! This would even map pretty well onto the [OpenAPI spec](https://swagger.io/specification/)'s `enum` property type.

**Comment by ajaishankar** [comment]
Looks like a _var_ needs a _type_ 😄 

In the DSL we could move the types to a separate section.

As long as we keep this minimal, the following could work...

```
vars:types {
  sort   : enum   start, -start, end, -end
  month  : Jan=1, Feb, Mar
  img    : file   format=binary
  images : file[] format=binary
  pic    : file   format=base64
}
```

In the UI we can add a type column for a var.

Please see discussion on how to specify [File Uploads](https://github.com/usebruno/bruno/issues/195) too

**Comment by svenluijten** [comment]
Instead of introducing a new `vars:types` block, I'd just keep it the same as it is now, with the type inferred (handled as it currently is) if no type is specified.

```
vars {
  message
  ~token
  sort  : enum(start, -start, end, -end)
  img   : file(format=binary)
}
```

**Comment by ajaishankar** [comment]
@svenluijten yes if we introduce a `vars:types` block that would define the type of a var.

If we want to keep this inline, feel it will be better for the type annotation to go on the left hand side, like the ~ in front of the token.

```
vars {
  message
  ~token
  @enum(start, -start, end, -end) sort :  -end
  @file(format=binary) img 
}
```
In the UI think we _might_ need a new optional column to denote the type (enum, file, file[]) etc

```
[name] [value] [type?]
```
Or when hovering over the variable name a popover link could be clicked to edit the type.

**Comment by helloanoop** [comment]
@ajaishankar

### Keeping TOML in mind
I would also consider our future plan to support TOML which currently seems to be a deccent possibility
Here is what chatgpt gave me when I asked how Bru would look in TOML: https://chat.openai.com/share/f92457d3-5c9c-4025-b369-cf7dc60a9b92

I think, we can only look at that post v1 on a more serious level, but I'd have it in back of my mind.
TOML supports enclosing [keys in quotes](https://github.com/toml-lang/toml/blob/main/toml.md#user-content-keys), I don't see a risk in introducing a special char at the key level

### DSL
I am almost in agreement, just running scenarios in my head if this would break anywhere
```groovy
vars {
  message: hello
  ~token: secret
  @disabled key: secret
  @enum(start, -start, end, -end) sort :  -end
  @file(format=binary) img 
}
```

PS: In the future, we can consider replacing `~` with `@disabled` too. Let me know what you think, I agree with the general sentiment that perhaps `~` is not as intuitive, and if we decided to support **directives**, why not making it common

### UI
We have a lot of people (>10k) using the app now, what I am seeing is that no-one actually is too concerned about the DSL. Everyone is pretty happy that the format is human readable (and would love to use a standard). Key point I am conveying is that very few people would hand-code the request on a file. If they really wanted to do that, they likely would choose a tool which is exclusively made for handcoding (https://hurl.dev/)

I also never hand-code, I just use the UI. Our UI should hide this abstraction neatly. Instead of providing a third column, we should provide a special row that gets rendered differently if the user has chosen a special var/param/header/path. Next to "+ Add Var" will be dropdown icon, and then the user can choose "+ Add Enum", "+ Add File"

I will put some small UI over the weekend to demo this idea. The file upload - https://github.com/usebruno/bruno/issues/195 is very crucial for the community and we need to get to it sooner than later

**Comment by helloanoop** [comment]
@conejoninja Happy to share that Bruno v0.24.0 supports request level documentation.

![image](https://github.com/usebruno/bruno/assets/9350733/6022f606-7f7b-491d-87e6-cf821ac13b59)

**Comment by Placeprom** [comment]
Hey @helloanoop , is there any progress on this feature? Even though the documentation enhances the experience adding the option for enum vars would be a huge improvement in DX. We currently solve this by creating a copy with each different option which clutters the collection.

**Comment by marartner** [comment]
Bumping to show support of this idea. Both for query params, headers, and even in bodies and such, having some sort of enum-support like this would be amazingly helpful.

**Comment by helloanoop** [comment]
We’re definitely on board to support this! We plan to take it up after completing [YAML support](https://github.com/usebruno/bruno/issues/3577), as it’s a prerequisite for the necessary schema updates.

Our rough ETA for this feature is January–February 2025.

**Comment by anusree-bruno** [comment]
> > Can you check this? I think this is closer to what you are suggesting [#344 (comment)](https://github.com/usebruno/bruno/issues/344#issuecomment-1747028765)
> 
> Yes, I think that would work. I'd suppose it's not limited to query params, but to all variables.

 #4014
